pallets/xyk-liquidity-mining/benchmarking/src/lib.rs

```diff
@@ -0,0 +1,461 @@
+
+		//global id: 5, yield id:6
+		lm_create_global_farm::<T>(100_000 * ONE, caller.clone(), Perquintill::from_percent(20))?;
+		lm_create_yield_farm::<T>(caller.clone(), 5, ASSET_PAIR, FixedU128::one())?;
+
+		//global id: 7, yield id:8
+		lm_create_global_farm::<T>(100_000 * ONE, caller.clone(), Perquintill::from_percent(20))?;
+		lm_create_yield_farm::<T>(caller.clone(), 7, ASSET_PAIR, FixedU128::one())?;
+
+		//global id: 9, yield id:10
+		lm_create_global_farm::<T>(100_000 * ONE, caller.clone(), Perquintill::from_percent(20))?;
+		lm_create_yield_farm::<T>(caller, 9, ASSET_PAIR, FixedU128::one())?;
+
+		set_block_number::<T>(200_000);
+
+		LiquidityMining::<T>::deposit_shares(RawOrigin::Signed(liq_provider.clone()).into(), GLOBAL_FARM_ID, YIELD_FARM_ID, ASSET_PAIR, shares_amount)?;
+		LiquidityMining::<T>::redeposit_lp_shares(RawOrigin::Signed(liq_provider.clone()).into(), GLOBAL_FARM_ID_2, YIELD_FARM_ID_2, ASSET_PAIR, DEPOSIT_ID)?;
+		LiquidityMining::<T>::redeposit_lp_shares(RawOrigin::Signed(liq_provider.clone()).into(), 5, 6, ASSET_PAIR, DEPOSIT_ID)?;
+		LiquidityMining::<T>::redeposit_lp_shares(RawOrigin::Signed(liq_provider.clone()).into(), 7, 8, ASSET_PAIR, DEPOSIT_ID)?;
+		LiquidityMining::<T>::redeposit_lp_shares(RawOrigin::Signed(liq_provider.clone()).into(), 9, 10, ASSET_PAIR, DEPOSIT_ID)?;
+
+		set_block_number::<T>(400_000);
+		let liq_provider_bsx_balance = MultiCurrencyOf::<T>::free_balance(BSX, &liq_provider);
+	}: {
+		LiquidityMining::<T>::claim_rewards(RawOrigin::Signed(liq_provider.clone()).into(), DEPOSIT_ID, 10)?
+	} verify {
+		assert!(MultiCurrencyOf::<T>::free_balance(BSX, &liq_provider).gt(&liq_provider_bsx_balance));
+	}
+
+	withdraw_shares {
+		let caller = create_funded_account::<T>("caller", 0);
+		let xyk_caller = create_funded_account::<T>("xyk_caller", 1);
+		let liq_provider = create_funded_account::<T>("liq_provider", 2);
+		let shares_amount = 10_000;
+
+		initialize_pool::<T>(xyk_caller, ASSET_PAIR.asset_in, ASSET_PAIR.asset_out, 1_000_000 * ONE, Price::from(10))?;
+		xyk_add_liquidity::<T>(liq_provider.clone(), ASSET_PAIR, shares_amount, 1_000_000_000)?;
+
+		//global id: 1, yield id: 2
+		lm_create_global_farm::<T>(100_000 * ONE, caller.clone(), Perquintill::from_percent(20))?;
+		lm_create_yield_farm::<T>(caller,GLOBAL_FARM_ID, ASSET_PAIR, FixedU128::one())?;
+
+		set_block_number::<T>(200_000);
+
+		LiquidityMining::<T>::deposit_shares(RawOrigin::Signed(liq_provider.clone()).into(), GLOBAL_FARM_ID, YIELD_FARM_ID, ASSET_PAIR, shares_amount)?;
+
+		set_block_number::<T>(400_000);
+		let liq_provider_bsx_balance = MultiCurrencyOf::<T>::free_balance(BSX, &liq_provider);
+	}: {
+		LiquidityMining::<T>::withdraw_shares(RawOrigin::Signed(liq_provider.clone()).into(), DEPOSIT_ID, YIELD_FARM_ID, ASSET_PAIR)?
+	} verify {
+		assert!(MultiCurrencyOf::<T>::free_balance(BSX, &liq_provider).gt(&liq_provider_bsx_balance));
+	}
```

Comment:
You probably want to trigger (and then check) the case where it is destroyed?

`withdraw_shares` is quite complex so I'm somewhat skeptical that you've covered the worst case with this short benchmark.

You might even want 2 or 3 for the different scenarios and then compare them in terms of their weight.

Reply:
yes, I think I compared case when deposit is destroyed and this one was worse. I'll double check and add comment.